civis/ml/_model.py

```diff
@@ -0,0 +1,443 @@
+"""Run CivisML jobs and retrieve the results
+"""
+import io
+import os
+import tempfile
+import threading
+import warnings
+from concurrent import futures
+from functools import wraps
+
+try:
+    import pandas as pd
+    HAS_PANDAS = True
+except ImportError:
+    HAS_PANDAS = False
+try:
+    import joblib
+    HAS_JOBLIB = True
+except ImportError:
+    HAS_JOBLIB = False
+
+from civis import APIClient, find_one
+from civis.base import CivisAPIError, CivisJobFailure
+import civis.io as cio
+from civis.futures import CivisFuture
+from civis.polling import _ResultPollingThread
+
+__all__ = ['ModelFuture', 'ModelError']
+
+
+class ModelError(RuntimeError):
+    def __init__(self, msg, estimator=None, metadata=None):
+        self.metadata = metadata
+        self.estimator = estimator
+        super().__init__(msg)
+
+
+def _check_is_fit(method):
+    """Makes sure that the ModelPipeline's been trained"""
+    @wraps(method)
+    def wrapper(*args, **kwargs):
+        self = args[0]
+        if not self.train_result_:
+            raise ValueError("This model hasn't been trained yet.")
+        return method(*args, **kwargs)
+    return wrapper
+
+
+def _block_and_handle_missing(method):
+    """For ModelFuture file-retrieving property methods.
+    Block until completion and attempt to retrieve result.
+    Raise exception only if the result isn't found.
+    """
+    @wraps(method)
+    def wrapper(self):
+        futures.wait((self,))  # Block until done
+        try:
+            return method(self)
+        except FileNotFoundError:
+            # We get here if the modeling job failed to produce
+            # any output and we don't have metadata.
+            if self.exception():
+                raise self.exception() from None
+            else:
+                raise
+    return wrapper
+
+
+def _stash_local_data(X, client=None):
+    """Store data in a temporary Civis File and return the file ID
+    """
+    civis_fname = 'modelpipeline_data.csv'
+    if HAS_PANDAS and isinstance(X, pd.DataFrame):
+        buf = io.BytesIO()
+        txt = io.TextIOWrapper(buf, encoding='utf-8')
+        X.to_csv(txt, encoding='utf-8', index=False)
+        txt.flush()
+        buf.seek(0)
+        file_id = cio.file_to_civis(buf, name=civis_fname, client=client)
+    else:
+        with open(X) as _fin:
+            file_id = cio.file_to_civis(_fin, name=civis_fname, client=client)
+
+    return file_id
+
+
+def _decode_train_run(train_job_id, train_run_id, client):
```

Comment:
I don't see where these private functions are being called: `_decode_train_run`, `_stash_local_data` and `_check_is_fit`.  Are they only used in testing?